### ISSUE TITLE
微调开始测试集acc非常低，单独eval，acc很高，原因是什么   ，使用PaddleOCR/configs/rec/PP-OCRv6 /PP-OCRv6_medium_rec.yml

### ISSUE THREAD

**Issue by hufeide** [PaddlePaddle/PaddleOCR#18281]
### 🔎 Search before asking

- [x] I have searched the PaddleOCR [Docs](https://paddlepaddle.github.io/PaddleOCR/) and found no similar bug report.
- [x] I have searched the PaddleOCR [Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) and found no similar bug report.
- [x] I have searched the PaddleOCR [Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions) and found no similar bug report.

### 🐛 Bug (问题描述)

微调开始测试集acc非常低，单独eval，acc很高，原因是什么   ，使用PaddleOCR/configs/rec/PP-OCRv6 /PP-OCRv6_medium_rec.yml

### 🏃‍♂️ Environment (运行环境)

python 3.13 
paddel ocr

### 🌰 Minimal Reproducible Example (最小可复现问题的Demo)

conda activate paddleocr
python3 -m paddle.distributed.launch --log_dir=./log/ --gpus '4'  tools/train.py -c /data/weidu_new/code_26/PaddleOCR/hxf/PP-OCRv6_medium_rec_fix.yml

python3 -m paddle.distributed.launch --log_dir=./log/ --gpus '4'  tools/eval.py -c /data/weidu_new/code_26/PaddleOCR/hxf/PP-OCRv6_medium_rec_fix.yml

**Comment by hufeide** [comment]
BatchNorm（BN）的统计量对模型输出有如此毁灭性的影响，本质是因为：**BN 并不是一个辅助模块，它直接参与了网络主干数据的线性变换。一旦统计量偏离，整层输出的分布就会被“拉扯”到全新的数值区间，导致后续所有卷积层和激活函数全部失效（激活值饱和或失效）。**

为了直观理解，我们用一个**极简的具体数值例子**拆解全过程：

---

## 简单数值示例：从 1.0 到 100.0 的分布漂移

假设我们关注网络中某一个特征图（Feature Map）在某个通道上的单个像素点 $x$。

BN 层的数学公式分为两步：

1. **标准化（Standardization）：**

$$\hat{x} = \frac{x - \mu}{\sqrt{\sigma^2 + \epsilon}}$$


2. **仿射变换（Scale and Shift）：**

$$y = \gamma \hat{x} + \beta$$



*(假设预训练好的参数为：$\gamma = 1.0$, $\beta = 0.0$)*

---

### 第一步：正常的预训练状态（Eval 得分高）

* **预训练阶段记录的真实统计量（通用大数据集）：**
* $\text{running\_mean} (\mu) = 100.0$
* $\text{running\_var} (\sigma^2) = 400.0 \implies \sigma = 20.0$



当输入的图像传入网络，算出的特征值为 **$x = 120.0$** 时：

1. **标准化计算：**

$$\hat{x} = \frac{120.0 - 100.0}{20.0} = 1.0$$


2. **BN 输出：**

$$y = 1.0 \times 1.0 + 0.0 = \mathbf{1.0}$$



> **后续激活函数（以 ReLU 为例）：**
> $\text{ReLU}(1.0) = \mathbf{1.0}$。特征正常保留并传递给下一层，模型推理完全准确（Eval Acc = 0.739）。

---

### 第二步：Training 阶段（发生污染）

现在模型进入 `train()` 模式，传入小 Batch 的手写体图片（假设手写体图片背景黑、线条细，整体像素值很低）。

在这个小 Batch 前向传播时，计算出的临时统计量为：

* $\mu_{\text{batch}} = 10.0$
* $\sigma_{\text{batch}} = 2.0$

由于 BN 的滑动更新机制：$\text{running\_mean} = (1 - m)\cdot\text{running\_mean} + m\cdot\mu_{\text{batch}}$，**只需跑几十个 Batch，`running_mean` 就会被快速拉低到大约 $15.0$**。

---

### 第三步：污染后重新 Eval（灾难发生）

再次切回 `eval()` 模式，模型使用**已经被污染的统计量**：

* 污染后的 $\text{running\_mean} (\mu_{\text{polluted}}) = \mathbf{15.0}$
* 污染后的 $\sigma_{\text{polluted}} = \mathbf{2.0}$

此时，如果传入同样的测试图像（特征值依然是 **$x = 120.0$**）：

1. **标准化计算：**

$$\hat{x}_{\text{polluted}} = \frac{120.0 - \mathbf{15.0}}{\mathbf{2.0}} = \mathbf{52.5}$$


2. **BN 输出：**

$$y = 1.0 \times 52.5 + 0.0 = \mathbf{52.5}$$



---

## 为什么这会导致 Acc 暴跌到 0.05？

原本预训练网络设计时，下一层卷积层期望收到的输入值是在 **`1.0`** 左右的合理数值。但现在由于 BN 的统计量污染，输出直接**放大到了 `52.5**`！

这种数值扭曲在深层网络中会产生极端的连锁反应：

1. **数值量级暴涨（或坍塌）**：经过多层卷积叠加后，特征图的数值要么爆炸（Inf/NaN），要么被拉到负数区。
2. **激活函数饱和/死亡**：
* 如果下一层是 **Sigmoid / Tanh**，输入 $52.5$ 会直接进入饱和区，输出变为恒定 $1.0$，梯度和差异性完全丧失。
* 如果输入变成了负数（比如 $-52.5$），经过 **ReLU** 后直接变成 $0$，导致大量神经元**永久死亡（Dying ReLU）**。


3. **分类头（Softmax）输出平坦**：传到最后一层 Softmax 时，输入的 Logits 数值完全错位，模型预测出的字符概率分布变成随机乱猜，Acc 直接跌到接近随机概率（0.05）。

---

## 总结

BN 的 `running_mean` 和 `running_var` **本质上是整个网络各层神经元的“坐标系网格”**。

污染 BN 统计量，相当于在不修改神经元权重（没有更新 Weight）的情况下，**把每一层神经元的参考坐标系整体强行平移和放大了几倍到几十倍**。后续所有的权重（Weight）都是基于旧坐标系训练出来的，自然无法识别被严重扭曲的新坐标系特征。

**Comment by hufeide** [comment]
可以在program.py加入以下参数，不更新BN                
‘’‘        for idx, batch in enumerate(train_dataloader):
            model.train()
                for layer in model.sublayers():
                    if 'BatchNorm' in layer.__class__.__name__:
                        layer.eval()、
‘’’

**Comment by crazy-wq** [comment]
> 可以在program.py加入以下参数，不更新BN ‘’‘ for idx, batch in enumerate(train_dataloader): model.train() for layer in model.sublayers(): if 'BatchNorm' in layer.**class**.**name**: layer.eval()、 ‘’’

是已经解决了？

**Comment by hufeide** [comment]
是的，解决了。



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2026年7月27日(周一) 上午10:42
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [PaddlePaddle/PaddleOCR] 微调开始测试集acc非常低，单独eval，acc很高，原因是什么   ，使用PaddleOCR/configs/rec/PP-OCRv6 /PP-OCRv6_medium_rec.yml (Issue #18281)


crazy-wq left a comment (PaddlePaddle/PaddleOCR#18281)
  
可以在program.py加入以下参数，不更新BN ‘’‘ for idx, batch in enumerate(train_dataloader): model.train() for layer in model.sublayers(): if 'BatchNorm' in layer.class.name: layer.eval()、 ‘’’
  
是已经解决了？
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!
